File: admin/python/test/test_db.py
"""
Unit tests for database module (db.py)
Tests connection pooling, URL parsing, and parameterized query execution
"""
import unittest
import os
import sys
from unittest.mock import Mock, MagicMock, patch

# Add parent directory to path to import modules
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

class TestDatabaseConnection(unittest.TestCase):
    """Test database connection and pool management"""
    
    def setUp(self):
        """Set up test fixtures"""
        self.test_db_url = "postgresql://testuser:testpass@localhost:5432/testdb"
        os.environ['DATABASE_URL'] = self.test_db_url
    
    def tearDown(self):
        """Clean up after tests"""
        if 'DATABASE_URL' in os.environ:
            del os.environ['DATABASE_URL']
    
    @patch('psycopg2.pool.ThreadedConnectionPool')
    def test_database_url_parsing(self, mock_pool):
        """Test DATABASE_URL environment variable parsing"""
        from db import DatabaseConnection
        
        # Create connection (should parse URL)
        db = DatabaseConnection()
        
        # Verify initialization was called with parsed components
        self.assertIsNotNone(db)
    
    @patch('psycopg2.pool.ThreadedConnectionPool')
    def test_connection_pool_creation(self, mock_pool):
        """Test that connection pool is created with correct parameters"""
        from db import DatabaseConnection
        
        mock_pool.return_value = MagicMock()
        db = DatabaseConnection()
        
        # Verify pool was created
        mock_pool.assert_called_once()
        args, kwargs = mock_pool.call_args
        
        # Check min and max connections (2-10 typical for ThreadedConnectionPool)
        self.assertEqual(args[0], 2)  # minconn
        self.assertEqual(args[1], 10)  # maxconn
    
    @patch('psycopg2.pool.ThreadedConnectionPool')
    def test_execute_query_parameterized(self, mock_pool):
        """Test parameterized query execution"""
        from db import DatabaseConnection
        
        # Mock connection and cursor
        mock_conn = MagicMock()
        mock_cursor = MagicMock()
        mock_cursor.fetchall.return_value = [{'id': 1, 'name': 'test'}]
        mock_cursor.description = [('id',), ('name',)]
        mock_conn.cursor.return_value.__enter__.return_value = mock_cursor
        
        mock_pool_instance = MagicMock()
        mock_pool_instance.getconn.return_value = mock_conn
        mock_pool.return_value = mock_pool_instance
        
        db = DatabaseConnection()
        
        # Execute query with parameters
        query = "SELECT * FROM users WHERE id = %s"
        params = (1,)
        result = db.execute_query(query, params)
        
        # Verify execute was called with query and params
        mock_cursor.execute.assert_called_once_with(query, params)
        self.assertIsNotNone(result)
    
    @patch('psycopg2.pool.ThreadedConnectionPool')
    def test_sql_injection_prevention(self, mock_pool):
        """Test that parameterized queries prevent SQL injection"""
        from db import DatabaseConnection
        
        mock_conn = MagicMock()
        mock_cursor = MagicMock()
        mock_cursor.fetchall.return_value = []
        mock_conn.cursor.return_value.__enter__.return_value = mock_cursor
        
        mock_pool_instance = MagicMock()
        mock_pool_instance.getconn.return_value = mock_conn
        mock_pool.return_value = mock_pool_instance
        
        db = DatabaseConnection()
        
        # Try query with potential SQL injection
        malicious_input = "1 OR 1=1"
        query = "SELECT * FROM users WHERE id = %s"
        
        db.execute_query(query, (malicious_input,))
        
        # Verify parameters were passed separately (not interpolated)
        call_args = mock_cursor.execute.call_args
        self.assertEqual(call_args[0][0], query)
        self.assertEqual(call_args[0][1], (malicious_input,))

class TestDatabaseIntegration(unittest.TestCase):
    """Integration tests with real database (if available)"""
    
    def setUp(self):
        """Check if test database is available"""
        self.db_available = os.getenv('TEST_DATABASE_URL') is not None
        if self.db_available:
            os.environ['DATABASE_URL'] = os.getenv('TEST_DATABASE_URL')
    
    @unittest.skipUnless(os.getenv('TEST_DATABASE_URL'), "Test database not configured")
    def test_real_connection(self):
        """Test actual database connection"""
        from db import DatabaseConnection
        
        db = DatabaseConnection()
        
        # Try simple query
        result = db.execute_query("SELECT 1 as test")
        self.assertIsNotNone(result)
        self.assertEqual(len(result), 1)

if __name__ == '__main__':
    unittest.main()

File: admin/cpp/test/test_db.cpp
/**
 * Unit tests for database module (db.cpp/db.h)
 * Tests connection pool, URL parsing, and parameterized query execution
 */

#include <gtest/gtest.h>
#include <gmock/gmock.h>
#include "../db.h"

using ::testing::_;
using ::testing::Return;

class DatabaseTest : public ::testing::Test {
protected:
    void SetUp() override {
        // Set test environment variable
        setenv("DATABASE_URL", "postgresql://testuser:testpass@localhost:5432/testdb", 1);
    }
    
    void TearDown() override {
        unsetenv("DATABASE_URL");
    }
};

TEST_F(DatabaseTest, DatabaseUrlParsing) {
    // Test that DatabaseConnection can parse DATABASE_URL
    EXPECT_NO_THROW({
        DatabaseConnection db;
    });
}

TEST_F(DatabaseTest, ConnectionPoolCreation) {
    // Test that connection pool is created
    DatabaseConnection db;
    
    // Connection pool should be initialized
    // (Implementation specific - may need to check internal state)
    EXPECT_TRUE(true); // Placeholder
}

TEST_F(DatabaseTest, ExecuteQueryWithParameters) {
    // Test parameterized query execution
    DatabaseConnection db;
    
    // Mock query execution
    std::string query = "SELECT * FROM users WHERE id = $1";
    std::vector<std::string> params = {"1"};
    
    // This would require actual database connection for full test
    // For unit test, we're testing the interface exists
    EXPECT_NO_THROW({
        // auto result = db.executeQuery(query, params);
    });
}

TEST_F(DatabaseTest, SqlInjectionPrevention) {
    // Test that parameterized queries prevent SQL injection
    DatabaseConnection db;
    
    std::string maliciousInput = "1' OR '1'='1";
    std::string query = "SELECT * FROM users WHERE id = $1";
    
    // With proper parameterization, this should be safe
    EXPECT_NO_THROW({
        // Parameters are passed separately, not interpolated
        std::vector<std::string> params = {maliciousInput};
        // auto result = db.executeQuery(query, params);
    });
}

TEST_F(DatabaseTest, ConnectionPoolThreadSafety) {
    // Test that connection pool is thread-safe
    DatabaseConnection db;
    
    // Multiple threads should be able to get connections
    std::vector<std::thread> threads;
    
    for (int i = 0; i < 5; ++i) {
        threads.emplace_back([&db]() {
            // Get connection from pool
            auto conn = db.getConnection();
            EXPECT_NE(conn, nullptr);
            
            // Simulate work
            std::this_thread::sleep_for(std::chrono::milliseconds(10));
            
            // Release connection
            db.releaseConnection(conn);
        });
    }
    
    for (auto& thread : threads) {
        thread.join();
    }
}

TEST_F(DatabaseTest, MaxConnectionsLimit) {
    // Test that connection pool respects max connections limit
    DatabaseConnection db;
    
    std::vector<std::shared_ptr<pqxx::connection>> connections;
    
    // Try to get more than max connections (10)
    for (int i = 0; i < 12; ++i) {
        try {
            auto conn = db.getConnection();
            if (conn) {
                connections.push_back(conn);
            }
        } catch (...) {
            // Expected to fail or block after 10 connections
        }
    }
    
    // Should have at most 10 connections
    EXPECT_LE(connections.size(), 10);
}

// Main function for running tests
int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
